File: Core/src/main/java/com/github/ezauton/core/action/DelayedAction.java
package com.github.ezauton.core.action;

import java.util.concurrent.TimeUnit;

/**
 * An action which ⏰ a certain amount of time before executing 1
 */
public class DelayedAction extends BaseAction {

    private Runnable runnable;
    private long millis;

    public DelayedAction(long value, TimeUnit unit) {
        millis = unit.toMillis(value);
    }

    public DelayedAction(long value, TimeUnit unit, Runnable runnable) {
        this(value, unit);
        this.runnable = runnable;
    }

    /**
     * Called when the time is up, i.e., the delay is done
     */
    private void onTimeUp() {
        if (runnable != null) {
            runnable.run();
        }
    }

    @Override
    public final void run(ActionRunInfo actionRunInfo) {
        try {
            actionRunInfo.getClock().sleep(millis, TimeUnit.MILLISECONDS);
        } catch (InterruptedException e) {
            return;
        }

        if (isStopped()) {
            return;
        }
        onTimeUp();
    }
}

File: Core/src/main/kotlin/com/github/ezauton/core/action/DelayedAction.kt
package com.github.ezauton.core.action

import com.github.ezauton.conversion.Time
import com.github.ezauton.core.action.require.Resource

/**
 * An action which ⏰ a certain amount of time before executing 1
 */
abstract class DelayedAction(waitDuration: Time, vararg val resources: Resource) : Action {

    private val waitMillis: Long = waitDuration.millis.toLong()

    /**
     * Called when the time is up, i.e., the delay is done
     */
    abstract fun onTimeUp()

    override suspend fun ActionContext.run() {
        try {
            delay(waitMillis)
        } catch (e: InterruptedException) {
            return
        }
        resources.map { it.take() }
        onTimeUp()
    }
}
